File: 프로그래머스/lv3/42627. 디스크 컨트롤러/디스크 컨트롤러.py
def solution(jobs):
    # jobs = [[0, 3], [1, 9], [2, 6]]
    
    total_time = 0
    min_start = 1001
    for i in range(len(jobs)):
        if jobs[i][0] < min_start:
            min_start = jobs[i][0]
    start = min_start
    len_jobs = len(jobs)
    
    while jobs:
    
        min_job = [0, 1001]   # [인덱스, 시간]
        for i in range(len(jobs)):
            
            if jobs[i][0] <= start and jobs[i][1] < min_job[1]:
                min_job = [i, jobs[i][1]]

        if min_job[1] != 1001:
            total_time += (start - jobs.pop(min_job[0])[0]) + min_job[1]
            start += min_job[1]
        else:
            min_start = 1001
            for i in range(len(jobs)):
                if jobs[i][0] < min_start:
                    min_start = jobs[i][0]
            start = min_start 

    answer = int(total_time / len_jobs)
    
    return answer

File: 프로그래머스/lv3/42627. 디스크 컨트롤러/디스크 컨트롤러.java
import java.util.PriorityQueue;
import java.util.ArrayList;
class Solution {
    public int solution(int[][] jobs) {
        PriorityQueue<Pair> pq = new PriorityQueue<Pair>();
        ArrayList<Pair> arr = new ArrayList<Pair>();
        int t = 1001;
        for (int[] j : jobs) {
            pq.add(new Pair(j[0], j[1]));
            t = Math.min(t, j[0]);
        }
        int answer = 0;
        while (!pq.isEmpty()) {
            Pair p = new Pair(0, 0);
            while (!pq.isEmpty()) {
                // 뽑고
                p = pq.poll();
                // 아니면 다시 넣기
                if (p.start > t) arr.add(p);
                else {
                    for (Pair a : arr) {
                        pq.add(a);
                    }
                    arr.clear();
                    break;
                };
                
                if (pq.isEmpty()) {
                    for (Pair a : arr) {
                        pq.add(a);
                    }
                    arr.clear();
                    t++;
                }
            }
            answer += (t - p.start) + p.time;
            // System.out.println(p.time);
            // System.out.println(answer);
            t += p.time;

        }
        answer /= jobs.length;
        return answer;
    }
}

class Pair implements Comparable<Pair> {
    int start;
    int time;
    
    public Pair(int start, int time) {
        this.start = start;
        this.time = time;
    }
    
    @Override
    public int compareTo(Pair other) {
        return Integer.compare(this.time, other.time);
    }
}